python/kfserving/kfserving/storage.py

```diff
@@ -67,10 +67,8 @@ def _download_s3(uri, temp_dir: str):
         bucket_path = bucket_args[1] if len(bucket_args) > 1 else ""
         objects = client.list_objects(bucket_name, prefix=bucket_path, recursive=True)
         for obj in objects:
-            # Replace any prefix from the object key with temp_dir
-            subdir_object_key = obj.object_name.replace(bucket_path, "", 1).strip("/")
             client.fget_object(bucket_name, obj.object_name,
-                               os.path.join(temp_dir, subdir_object_key))
+                               os.path.join(temp_dir, obj.object_name.split("/")[-1]))
```

Comment:
Hmm - can you take a look at GCS and Azure implementations... the following cases need to be handled:
```
some/src/path -> /temp_dir/<contents of path>
some/src/path/ -> /temp_dir/<contents of path>
some/src/path/foo.model -> /temp_dir/foo.model
/ -> /temp_dir/<contents of />
```

Note that we trim the path prefix so data is copied directly to `temp_dir` not to `temp_dir/some/src/path/`

Reply:
Which this change, is all the object files will be under the same directory? 

Probably we may want to keep the old way of pulling the objects as well because some Serving framework such as TensorFlow Serving may assume certain directory structure. 

For the TensorFlow example, we are loading the model in the SavedModel format which it assumes the model files and variables is under a certain directory. (e.g. the first model will be under dir `0001` and all the variables are under dir `variables`.)

Reply:
so this is how it needs to work:

Given source bucket has:
```
some/src/path/foo.config
some/src/path/1/model.pb
```
If `modelUri=some/src/path`

Destination should look like:
```
temp_dir/foo.config
temp_dir/1/model.pb
```

So folder structure under `some/src/path` is replicated

Reply:
Sadly the unit tests are way lacking here (mostly my fault) :-(

Reply:
> Hmm - can you take a look at GCS and Azure implementations...
Unfortunately not, I don't have any creds for GCS and Azure to startwith..

> Destination should look like:
> temp_dir/1/model.pb
I see, didn't think of this case, will fix. Thanks.

Reply:
> I don't have any creds for GCS and Azure to startwith

I meant you can look at the implementations in storage.py :)

Reply:
Oh sorry, so for GCS and Azure this has been supported ?

Reply:
see #309 (seems there were some bugs!)